### PR TITLE
drivers: rcar_mmc: handle unaligned access to data buffer

### PR DESCRIPTION
**Pull request by xakep-amatop**

### CODE REVIEW

File: drivers/sdhc/rcar_mmc.c [hunk 1]
```diff
@@ -657,11 +716,11 @@ static int rcar_mmc_sd_buf_rx_tx_data(const struct device *dev,
 			uint8_t copy_size = MIN(sd_buf0_size, data->block_size - w_off);
 
 			if (is_read) {
-				buf0 = sys_read64(DEVICE_MMIO_GET(dev) + RCAR_MMC_BUF0);
+				buf0 = rcar_mmc_read_buf0(dev);
```
Comment (firscity) [inline drivers/sdhc/rcar_mmc.c:790]: Just out of curiosity, why this patches are not squashed?

File: drivers/sdhc/rcar_mmc.c [hunk 2]
```diff
@@ -659,11 +652,16 @@ static int rcar_mmc_sd_buf_rx_tx_data(const struct device *dev,
 		info2_reg &= ~info2_poll_flag;
 		rcar_mmc_write_reg32(dev, RCAR_MMC_INFO2, info2_reg);
 
-		for (w_off = 0; (w_off * sizeof(*buf)) < data->block_size; w_off++) {
+		for (w_off = 0; w_off < aligned_block_size; w_off += sd_buf0_size) {
+			uint64_t buf0 = 0ULL;
```
Comment (lorc) [inline drivers/sdhc/rcar_mmc.c:786]: I am curious what will happen with the remaining of the buffer. For example, we have only 1 block with block size = 1. Your code will send 8 bytes: 1 byte of data and 7 zeroes. Will those 7 zeroes overwrite data past the intended block?

Comment (xakep-amatop) [reply drivers/sdhc/rcar_mmc.c:786]: unfortunately, according to data from CSD structure (H3ULCB board), I can't use block size isn't equal to 512:
`READ_BL_LEN [83:80] == 9` and `READ_BL_PARTIAL [79] == 0` and `READ_BLK_MISALIGN [77] == 0`
`WRITE_BL_LEN [25:22] == 9` and `WRITE_BL_PARTIAL[21] == 0` and `WRITE_BLK_MISALIGN [78] == 0`

```
The data block length is computed as 2 ^ READ_BL_LEN for reading and 2 ^ WRITE_BL_LEN for writing.
```

```
WRITE_BL_PARTIAL defines whether partial block sizes can be used in block write commands.
WRITE_BL_PARTIAL=’0’ means that only the 512B and the WRITE_BL_LEN block size can
be used for block oriented data write.
WRITE_BL_PARTIAL=’1’ means that smaller blocks can be used as well. The minimum block
size is one byte.
Higher than 2 GB of density (sector access mode):
WRITE_BL_PARTIAL=’0’ means that only the 512B and the WRITE_BL_LEN block size can
be used for block oriented data write.
WRITE_BL_PARTIAL=’1’ means that smaller blocks can be used as well. The minimum block
size will be equal to minimum addressable unit, one sector (512B).
```
the same description about `READ_BL_PARTIAL`.


```
7.3.9 WRITE_BLK_MISALIGN [78]
Defines if the data block to be written by one command can be spread over more than one physical block
of the memory device. The size of the memory block is defined in WRITE_BL_LEN.
WRITE_BLK_MISALIGN=0 signals that crossing physical block boundaries is invalid.
WRITE_BLK_MISALIGN=1 signals that crossing physical block boundaries is allowed.
```

Moreover, Zephyr uses 512 block size in all places where we need to know the block size

Comment (xakep-amatop) [reply drivers/sdhc/rcar_mmc.c:786]: for testing the case mentioned by you, it is needed another SD or MMC with appropriate capabilities: 1-byte block writing support

File: drivers/sdhc/rcar_mmc.c (hunk 1 above)
Comment (xakep-amatop) [reply drivers/sdhc/rcar_mmc.c:790]: at the first time, they were, but I've split one commit into two: one for handling unaligned access and another for working with different sd_buf0 access width sizes

if it is needed, I can combine them with one another

File: drivers/sdhc/rcar_mmc.c (hunk 2 above)
Comment (lorc) [reply drivers/sdhc/rcar_mmc.c:786]: Okay, so if we can't test this code, how we can be sure that is working properly? Just by looking at it, I can assume that it should work fine at least for block sizes >= 8 bytes. 

I believe you need to add additional check to make sure that block size at least 8 bytes, along with a comment that we can't be sure that smaller block sizes would work at all.

Comment (xakep-amatop) [reply drivers/sdhc/rcar_mmc.c:786]: restore prev checks and add extra checks of block size in a new commit

**Review by lorc — COMMENTED** [review]
`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

**Review by firscity — APPROVED** [review]
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

**Pull request merged**